### ISSUE TITLE
meta: remaining MidiDeviceService tasks

### ISSUE THREAD

**Issue by atsushieno** [atsushieno/android-audio-plugin-framework#82]
(update: see https://github.com/atsushieno/android-audio-plugin-framework/issues/82#issuecomment-892459550 which updates all the statuses involved)

(1) At this state AAPMidiProcessor can process only MIDI 2.0 messages. For MIDI 1.0 messages we have to sort out how to adjust lengths in the messages along with modified timestamps per audio processing cycles. MIDI 1.0 is especially problematic because length bytes can vary. We likely have to change how MIDI 1.0 messages are structured on the existing channels.

(1.1) Maybe we should simply drop it and process everything in MIDI 2.0 UMPs. Then every subprojects need to reflect the protocol changes. aap-juce and all the plugins, plugin hosts, aap-lv2 based plugins (relatively easy as we already have UMP processing implementation)

(2) At this state we assume that an AAP application contains at most one instrument plugin. It is already wrong in aap-lv2-mda as there are handful of instruments (EPiano, DX10, JX10, Piano...) in the app. A possible solution would be to create `<input>` port for every instrument. Then we only have to (a) create MidiReceiver instances for each, and (b) determine how to map those input entries to MidiReceivers (stable indices suffice).

(3) there are some FIXMEs on terminating /unbinding device service.

**Comment by atsushieno** [comment]
(3) unlike aap-ayumi, aap-fluidsynth does not work as expected. Needs investigation.

**Comment by atsushieno** [comment]
(4) aap-ayumi fails to instantiate MidiReceiver if it was not started via Activity. Some initialization step is missing from Service.

**Comment by atsushieno** [comment]
re.(3) - `fluidsynth_lv2_run()` receives MIDI 1.0 events from kmmk (either in MIDI 1.0 or MIDI 2.0). There is likely some problem on resolving sf3 assets.

**Comment by atsushieno** [comment]
(5) the overall output audio quality is horrible. Lots of noises involved.

re. (4) - a good news is that I could confirm not only on aap-ayumi - now aap-sfizz can generate valid outputs (ignoring the noise issue), IF it is MIDI 2.0 (on kmmk). MIDI 1.0 issue needs investigation.

**Comment by atsushieno** [comment]
aap-juce-obxd got a working MidiDeviceService now (also suffered from the same kind of noise).

**Comment by atsushieno** [comment]
re. (6) - it ssems that if we set Oboe audio output channels to 1 then it's generating correct-ish output (still some glitches).

re. (4) - fluidsynth has problem on building native shared libraries. And it never worked on arm64.

re. (3) - `AAPMidiProcessor::terminate()` crashes at `close()` on random FDs triggered by fdsan.

**Comment by atsushieno** [comment]
(6) is fixed at 1ca9df7.

**Comment by atsushieno** [comment]
re. (4) - With the latest build system overhaul, aap-fluidsynth is back with working arm64 support. But now it's not working with aap-midi-device-service app or via kmmk. Needs investigation again.

**Comment by atsushieno** [comment]
re. (4) - aap-fluidsynth still has problem generating audio outputs, aap-sfizz generates audio outputs. The output is still awkward, either the frequency is weird or contains glitchy noise.

**Comment by atsushieno** [comment]
re. (1) - it is blocked by somewhat awkward reason. I use [kmmk](https://github.com/atsushieno/kmmk) as a MIDI client for dogfooding MIDI 2.0 support, but it does not build on Android on IntelliJ IDEA 2021.1.2EAP (neither on 2021.1.1). IDEA thinks its Android plugin is not up to date enough, which is weird in any sense (AGP, not Android plugin, is 4.2.1, and Android plugin is 10.4.1.1.*).

> This version of the Android Support plugin for IntelliJ IDEA (or Android Studio) cannot open this project, please retry with version 4.2 or newer

![image](https://user-images.githubusercontent.com/53929/122879647-1a8c4700-d374-11eb-85f4-3c94cc186177.png)

Technically it is still possible to open, build and run in Android Studio, it does not resolve any MPP dependencies and thus coding experience is awkward.

**Comment by atsushieno** [comment]
There has been various changes that affect here.

- There is no build blocker. Bumbeblee after alpha04 works great.
- Fluidsynth and aap-fluidsynth builds with no problem nowadays. I also revived [fluidsynth-midi-service-j](https://github.com/atsushieno/fluidsynth-midi-service-j) that ensures that fluidsynth on Android works in general.
- Regarding (1) MIDI protocols, I'm changing my mind to support both MIDI 1.0 and 2.0 protocols and leave them handled by the recipient. It is because -
  - (a) `juceaap_AAPWrappers.cpp` (applied to OB-Xd for example)  still accepts only MIDI 1.0 ATM. It can implement MIDI 2.0 too but supporting MIDI 1.0 in `AAPMidiProcessor.cpp` is easier.
  - (b) I have been making changes to MIDI 2.0 support in various ktmidi bits (for example kmmk) to rather use existing MIDI 1.0 channel and promote to MIDI 2.0 by MIDI-CI Set New Protocol, *dynamically* (at run time) i.e. we process MIDI 1.0 messages on the same connection too.
- Since we have made changes to AAPMidiProcessor to process only MIDI 2.0, JUCE synths do not work with StandaloneAudioPluginMidiDeviceService at this state (if updated to the latest AAP).
- Regarding (2) multi-port plugins, it is addressed at fb48bba (not sure if it fixed the issue, needs to verify with aap-mda-lv2).
- Regarding (3) i.e. AAPMidiProcessor::terminate() crashes at close() on random FDs triggered by fdsan - maybe it still exists.

AAPMidiProcessor (in androidaudioplugin-midi-device-service) can now accept MIDI-CI Set New Protocol (as of c588033) to switch to MIDI 2.0 when it is receiving MIDI messages. On the other hand, it generates and sends only (at this state) MIDI 2.0 messages to the AAP instrument instance (and thus JUCEAAPWrappers does not understand it).

**Comment by atsushieno** [comment]
More on (1) buffering MIDI 1.0 messages at `AAPMidiProcessor` before sending them to the plugin:

Currently AAP "midi" channels expect the data at `process()` in the format below:

- 4 bytes of Time Unit specification (same as SMF)
- 4 bytes of buffer size.
- MIDI buffer (take it as SMF Format 0 track)

In MIDI 2 mode, it expects

- 4 bytes of (reserved)
- 4 bytes of buffer size
- 24 bytes of (reserved)
- MIDI buffers as in UMPs

Unlike stated at the issue description, MIDI1 buffering is not really problematic as lengths in MIDI events would not change in the AAP buffers, and the total length is of a fixed 32-bit integer that does not change.

**Comment by atsushieno** [comment]
aap-juce-obxd is working again with MIDI 1.0 protocol, can be connected from kmmk and any MIDI client apps (e.g. N-Track DAW).

**Comment by atsushieno** [comment]
Back to this project. Currently audio outputs are weird. It seems to work if Oboe channel configuration (`setChannelCount()`) is mono, meaning that our output buffer is likely wrong.

**Comment by atsushieno** [comment]
469d96a adds a workaround for ^.

**Comment by atsushieno** [comment]
Re: (2), more than one MDA-LV2 instruments are listed as expected, but the instantiated plugin seems to be only the identical one (DX10?).

**Comment by atsushieno** [comment]
Now that both aap-lv2 and aap-juce are confirmed "kinda" working with the Oboe channels workaround (469d96a), we should close this meta issue and create a separate issue for (2). Maybe for (3) if it still exists.

**Comment by atsushieno** [comment]
closing (finally!)
